flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -21,519 +21,897 @@
-	 * or really rejected by TaskManager. We shall retry this request by:
-	 * <ul>
-	 * <li>1. verify and clear all the previous allocate information for this request
-	 * <li>2. try to request slot again
-	 * </ul>
-	 * <p>
-	 * This may cause some duplicate allocation, e.g. the slot request to TaskManager is successful but the response
-	 * is lost somehow, so we may request a slot in another TaskManager, this causes two slots assigned to one request,
-	 * but it can be taken care of by rejecting registration at JobManager.
+	 * Reports the current slot allocations for a task manager identified by the given instance id.
 	 *
-	 * @param originalRequest The original slot request
-	 * @param slotId          The target SlotID
+	 * @param instanceId identifying the task manager for which to report the slot status
+	 * @param slotReport containing the status for all of its slots
+	 * @return true if the slot status has been updated successfully, otherwise false
 	 */
-	void handleSlotRequestFailedAtTaskManager(final SlotRequest originalRequest, final SlotID slotId) {
-		final AllocationID originalAllocationId = originalRequest.getAllocationId();
-		LOG.info("Slot request failed at TaskManager, SlotID:{}, AllocationID:{}, JobID:{}",
-			slotId, originalAllocationId, originalRequest.getJobId());
-
-		if (allocationMap.isAllocated(slotId)) {
-			final AllocationID expectedAllocationId = allocationMap.getAllocationID(slotId);
-
-			// check whether we have an agreement on whom this slot belongs to
-			if (originalAllocationId.equals(expectedAllocationId)) {
-				LOG.info("De-allocate this request and retry");
-				allocationMap.removeAllocation(expectedAllocationId);
-				pendingSlotRequests.put(originalRequest.getAllocationId(), originalRequest);
-				ResourceSlot slot = checkNotNull(getRegisteredSlot(slotId));
-				// treat this slot as empty and retry with a different request
-				handleFreeSlot(slot);
+	public boolean reportSlotStatus(InstanceID instanceId, SlotReport slotReport) {
+		checkInit();
+
+		TaskManagerRegistration taskManagerRegistration = taskManagerRegistrations.get(instanceId);
+
+		if (null != taskManagerRegistration) {
+			HashSet<SlotID> slotsToRemove = new HashSet<>(taskManagerRegistration.getSlots());
+			boolean idle = true;
+
+			for (SlotStatus slotStatus : slotReport) {
+				if (slotsToRemove.remove(slotStatus.getSlotID())) {
+					// slot which was already registered
+					updateSlot(slotStatus.getSlotID(), slotStatus.getAllocationID());
+				} else {
+					// new slot
+					registerSlot(
+						instanceId,
+						slotStatus.getSlotID(),
+						slotStatus.getAllocationID(),
+						slotStatus.getResourceProfile(),
+						taskManagerRegistration.getTaskManagerConnection());
+				}
+
+				TaskManagerSlot slot = slots.get(slotStatus.getSlotID());
+
+				idle &= slot.isFree();
+			}
+
+			// remove the slots for which we haven't received a slot status message
+			removeSlots(slotsToRemove);
+
+			if (idle) {
```

Comment:
I assume this breaks the TaskManager timeouts. Every time a heartbeat comes (with a slot report), the TaskManager is detected to be idle, and the timeout is scheduled, overriding the previous timeout (pushing the timeout further back).